File: watermark.rb
#!/usr/bin/env ruby
# watermark.rb
# encoding: UTF-8

require 'rmagick'
require 'optparse'
require 'fileutils'

COLORS = {
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

POSITIONS = {
  'tl' => [0.05, 0.05],
  'tc' => [0.5, 0.05],
  'tr' => [0.95, 0.05],
  'ml' => [0.05, 0.5],
  'c'  => [0.5, 0.5],
  'mr' => [0.95, 0.5],
  'bl' => [0.05, 0.95],
  'bc' => [0.5, 0.95],
  'br' => [0.95, 0.95]
}

def get_position(img, pos_key)
  w, h = img.columns, img.rows
  x_ratio, y_ratio = POSITIONS[pos_key] || POSITIONS['br']
  [(w * x_ratio).to_i, (h * y_ratio).to_i]
end

def add_text_watermark(input_path, output_path, text, position: 'br',
                       opacity: 0.5, font_size: 48, color: '#FFFFFF',
                       rotate: 0)
  img = Magick::Image.read(input_path).first
  w, h = img.columns, img.rows

  # Создаём слой для текста
  overlay = Magick::Image.new(w, h) { self.background_color = 'none' }
  draw = Magick::Draw.new

  # Настройка шрифта
  draw.font_family = 'Arial'
  draw.font_size = font_size
  draw.fill = color
  draw.gravity = Magick::CenterGravity

  # Позиция
  x, y = get_position(img, position)
  if position == 'c'
    x = w / 2
    y = h / 2
    draw.gravity = Magick::CenterGravity
  else
    draw.gravity = Magick::NorthWestGravity
    if position.include?('r')
      draw.gravity = Magick::NorthEastGravity
      x = w - x
    end
    if position.include?('b')
      draw.gravity = Magick::SouthWestGravity
      y = h - y
    end
    if position.include?('r') && position.include?('b')
      draw.gravity = Magick::SouthEastGravity
    end
  end

  draw.annotate(overlay, 0, 0, x, y, text)

  # Применяем прозрачность
  overlay.alpha(Magick::ActivateAlphaChannel)
  if opacity < 1.0
    overlay.each_pixel do |pixel, col, row|
      pixel.alpha = (pixel.alpha * opacity).to_i
    end
  end

  # Поворот
  if rotate != 0
    overlay = overlay.rotate(rotate)
  end

  # Накладываем
  result = img.composite(overlay, Magick::CenterGravity, Magick::OverCompositeOp)
  result.write(output_path)
end

def add_image_watermark(input_path, output_path, watermark_path, position: 'br',
                        opacity: 0.5, rotate: 0)
  img = Magick::Image.read(input_path).first
  wm = Magick::Image.read(watermark_path).first

  # Масштабируем
  if wm.columns > img.columns / 2 || wm.rows > img.rows / 2
    scale = [img.columns.to_f / wm.columns, img.rows.to_f / wm.rows].min * 0.4
    wm.resize!( (wm.columns * scale).to_i, (wm.rows * scale).to_i )
  end

  # Применяем прозрачность
  if opacity < 1.0
    wm.alpha(Magick::ActivateAlphaChannel)
    wm.each_pixel do |pixel, col, row|
      pixel.alpha = (pixel.alpha * opacity).to_i
    end
  end

  # Позиция
  x, y = get_position(img, position)
  if position == 'c'
    x = (img.columns - wm.columns) / 2
    y = (img.rows - wm.rows) / 2
  else
    if position.include?('r')
      x = img.columns - wm.columns - x
    end
    if position.include?('b')
      y = img.rows - wm.rows - y
    end
  end

  # Накладываем
  result = img.composite(wm, x, y, Magick::OverCompositeOp)
  result.write(output_path)
end

def batch_process(input_dir, output_dir, mode, **kwargs)
  FileUtils.mkdir_p(output_dir)

  extensions = %w[.jpg .jpeg .png .bmp .tiff .webp]
  files = Dir.entries(input_dir)
             .select { |f| extensions.include?(File.extname(f).downcase) }

  files.each_with_index do |fname, i|
    in_path = File.join(input_dir, fname)
    out_path = File.join(output_dir, fname)
    puts colorize("[#{i+1}/#{files.length}] Processing #{fname}...", :yellow)

    begin
      if mode == 'text'
        add_text_watermark(in_path, out_path, kwargs[:text],
                           position: kwargs[:position],
                           opacity: kwargs[:opacity],
                           font_size: kwargs[:font_size],
                           color: kwargs[:color],
                           rotate: kwargs[:rotate])
      else
        add_image_watermark(in_path, out_path, kwargs[:watermark],
                            position: kwargs[:position],
                            opacity: kwargs[:opacity],
                            rotate: kwargs[:rotate])
      end
    rescue => e
      puts colorize("  Error: #{e.message}", :red)
    end
  end
end

options = { position: 'br', opacity: 0.5, font_size: 48, color: '#FFFFFF', rotate: 0 }
parser = OptionParser.new do |opts|
  opts.banner = "Usage: ruby watermark.rb text|image|batch <input> [output] [options]"
  opts.on("-t", "--text TEXT", "Watermark text") { |v| options[:text] = v }
  opts.on("-i", "--image FILE", "Watermark image") { |v| options[:watermark] = v }
  opts.on("-p", "--position POS", "Position: tl,tc,tr,ml,c,mr,bl,bc,br (default: br)") { |v| options[:position] = v }
  opts.on("-o", "--opacity N", Float, "Opacity 0.0-1.0 (default: 0.5)") { |v| options[:opacity] = v }
  opts.on("-s", "--size N", Integer, "Font size (default: 48)") { |v| options[:font_size] = v }
  opts.on("-c", "--color COLOR", "Text color #RRGGBB (default: #FFFFFF)") { |v| options[:color] = v }
  opts.on("-r", "--rotate DEG", Integer, "Rotation angle (default: 0)") { |v| options[:rotate] = v }
end
parser.parse!

args = ARGV
if args.length < 2
  puts colorize(parser.banner, :yellow)
  exit 1
end

mode = args[0]
input = args[1]
output = args[2] || "watermarked_#{File.basename(input)}"

begin
  if mode == 'batch'
    if output.nil? || File.directory?(input) == false
      raise "Input must be a directory for batch mode"
    end
    batch_process(input, output, mode, **options)
    puts colorize("Batch processing completed!", :green)
  elsif mode == 'text'
    if options[:text].nil? || options[:text].empty?
      raise "Text required for text mode"
    end
    add_text_watermark(input, output, options[:text],
                       position: options[:position],
                       opacity: options[:opacity],
                       font_size: options[:font_size],
                       color: options[:color],
                       rotate: options[:rotate])
    puts colorize("Done! Result saved to #{output}", :green)
  elsif mode == 'image'
    if options[:watermark].nil? || options[:watermark].empty?
      raise "Watermark image required for image mode"
    end
    add_image_watermark(input, output, options[:watermark],
                        position: options[:position],
                        opacity: options[:opacity],
                        rotate: options[:rotate])
    puts colorize("Done! Result saved to #{output}", :green)
  else
    raise "Invalid mode. Use text, image, or batch"
  end
rescue => e
  puts colorize("Error: #{e.message}", :red)
  exit 1
end

File: watermark.js
// watermark.js
#!/usr/bin/env node
'use strict';

const sharp = require('sharp');
const fs = require('fs');
const path = require('path');

const COLORS = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const POSITIONS = {
    tl: { x: 0.05, y: 0.05 },
    tc: { x: 0.5, y: 0.05 },
    tr: { x: 0.95, y: 0.05 },
    ml: { x: 0.05, y: 0.5 },
    c:  { x: 0.5, y: 0.5 },
    mr: { x: 0.95, y: 0.5 },
    bl: { x: 0.05, y: 0.95 },
    bc: { x: 0.5, y: 0.95 },
    br: { x: 0.95, y: 0.95 }
};

async function addTextWatermark(inputPath, outputPath, text, position = 'br',
                                 opacity = 0.5, fontSize = 48, color = '#FFFFFF',
                                 rotate = 0) {
    const img = sharp(inputPath);
    const metadata = await img.metadata();

    // Создаём SVG с текстом
    const svg = `
        <svg width="${metadata.width}" height="${metadata.height}">
            <defs>
                <style>
                    .text { font-size: ${fontSize}px; font-family: sans-serif; fill: ${color};
                            opacity: ${opacity}; }
                </style>
            </defs>
            <text class="text" x="${POSITIONS[position].x * metadata.width}"
                  y="${POSITIONS[position].y * metadata.height}"
                  text-anchor="${position.includes('r') ? 'end' : position.includes('l') ? 'start' : 'middle'}"
                  dominant-baseline="${position.includes('b') ? 'baseline' : position.includes('t') ? 'hanging' : 'central'}">
                ${text}
            </text>
        </svg>
    `;

    const overlay = Buffer.from(svg);

    await img
        .composite([{ input: overlay, blend: 'over' }])
        .toFile(outputPath);
}

async function addImageWatermark(inputPath, outputPath, watermarkPath, position = 'br',
                                  opacity = 0.5, rotate = 0) {
    const img = sharp(inputPath);
    const metadata = await img.metadata();
    const wm = sharp(watermarkPath);
    const wmMeta = await wm.metadata();

    // Масштабируем водяной знак
    let wmW = wmMeta.width, wmH = wmMeta.height;
    const maxW = metadata.width * 0.4, maxH = metadata.height * 0.4;
    if (wmW > maxW || wmH > maxH) {
        const ratio = Math.min(maxW / wmW, maxH / wmH);
        wmW = Math.floor(wmW * ratio);
        wmH = Math.floor(wmH * ratio);
    }

    let x = POSITIONS[position].x * metadata.width;
    let y = POSITIONS[position].y * metadata.height;
    if (position === 'c') {
        x = (metadata.width - wmW) / 2;
        y = (metadata.height - wmH) / 2;
    } else {
        if (position.includes('r')) x -= wmW;
        if (position.includes('b')) y -= wmH;
        if (position.includes('l')) x += 0;
        if (position.includes('t')) y += 0;
    }

    const wmBuffer = await wm
        .resize(wmW, wmH)
        .ensureAlpha()
        .raw()
        .toBuffer({ resolveWithObject: true });

    // Применяем прозрачность
    const data = wmBuffer.data;
    for (let i = 3; i < data.length; i += 4) {
        data[i] = Math.round(data[i] * opacity);
    }

    const wmWithAlpha = sharp(data, {
        raw: { width: wmW, height: wmH, channels: 4 }
    });

    const wmPng = await wmWithAlpha.png().toBuffer();

    await img
        .composite([{ input: wmPng, left: Math.round(x), top: Math.round(y) }])
        .toFile(outputPath);
}

async function batchProcess(inputDir, outputDir, mode, options) {
    if (!fs.existsSync(outputDir)) fs.mkdirSync(outputDir, { recursive: true });

    const extensions = ['.jpg', '.jpeg', '.png', '.bmp', '.tiff', '.webp'];
    const files = fs.readdirSync(inputDir)
        .filter(f => extensions.includes(path.extname(f).toLowerCase()));

    for (let i = 0; i < files.length; i++) {
        const inPath = path.join(inputDir, files[i]);
        const outPath = path.join(outputDir, files[i]);
        console.log(colorize(`[${i+1}/${files.length}] Processing ${files[i]}...`, 'yellow'));

        try {
            if (mode === 'text') {
                await addTextWatermark(inPath, outPath, options.text, options.position,
                                       options.opacity, options.size, options.color, options.rotate);
            } else {
                await addImageWatermark(inPath, outPath, options.image, options.position,
                                        options.opacity, options.rotate);
            }
        } catch (err) {
            console.log(colorize(`  Error: ${err.message}`, 'red'));
        }
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length < 2) {
        console.log(colorize('Usage: node watermark.js text|image|batch <input> [output] [options]', 'yellow'));
        console.log('Options:');
        console.log('  -t, --text <text>       Watermark text');
        console.log('  -i, --image <file>      Watermark image');
        console.log('  -p, --position <pos>    Position: tl,tc,tr,ml,c,mr,bl,bc,br (default: br)');
        console.log('  -o, --opacity <n>       Opacity 0.0-1.0 (default: 0.5)');
        console.log('  -s, --size <n>          Font size (default: 48)');
        console.log('  -c, --color <color>     Text color #RRGGBB (default: #FFFFFF)');
        console.log('  -r, --rotate <deg>      Rotation angle (default: 0)');
        process.exit(1);
    }

    const mode = args[0];
    const input = args[1];
    let output = args[2] || (mode === 'batch' ? null : `watermarked_${path.basename(input)}`);

    const options = { position: 'br', opacity: 0.5, size: 48, color: '#FFFFFF', rotate: 0 };
    for (let i = 0; i < args.length; i++) {
        switch (args[i]) {
            case '-t': case '--text': options.text = args[++i]; break;
            case '-i': case '--image': options.image = args[++i]; break;
            case '-p': case '--position': options.position = args[++i]; break;
            case '-o': case '--opacity': options.opacity = parseFloat(args[++i]); break;
            case '-s': case '--size': options.size = parseInt(args[++i]); break;
            case '-c': case '--color': options.color = args[++i]; break;
            case '-r': case '--rotate': options.rotate = parseInt(args[++i]); break;
        }
    }

    try {
        if (mode === 'batch') {
            if (!output) throw new Error('Output directory required for batch mode');
            await batchProcess(input, output, mode, options);
            console.log(colorize('Batch processing completed!', 'green'));
        } else if (mode === 'text') {
            if (!options.text) throw new Error('Text required for text mode');
            await addTextWatermark(input, output, options.text, options.position,
                                   options.opacity, options.size, options.color, options.rotate);
            console.log(colorize(`Done! Result saved to ${output}`, 'green'));
        } else if (mode === 'image') {
            if (!options.image) throw new Error('Watermark image required for image mode');
            await addImageWatermark(input, output, options.image, options.position,
                                    options.opacity, options.rotate);
            console.log(colorize(`Done! Result saved to ${output}`, 'green'));
        } else {
            throw new Error('Invalid mode. Use text, image, or batch');
        }
    } catch (err) {
        console.log(colorize(`Error: ${err.message}`, 'red'));
        process.exit(1);
    }
}

main();
